README.md
# Prometheus Microovn Exporter

## A simple snap package that exports minimal data from a ovn cluster such as:
1. Cluster status
2. Open ports
3. Valid certs

*Can use with either microovn or ovn (chassis and central both)*

Example output (here, 0 generally means good, as in cluster state is good, port is open, cert is valid)

```console
$ curl localhost:9999
...
# HELP ovn_cluster_state State of ovn ovn_cluster_state
# TYPE ovn_cluster_state gauge
ovn_cluster_state{cluster_status="nb"} 0.0
ovn_cluster_state{cluster_status="sb"} 0.0
# HELP ovn_cluster_state State of ovn ovn_cluster_state
# TYPE ovn_cluster_state gauge
ovn_cluster_state{ports="6641"} 0.0
ovn_cluster_state{ports="6642"} 0.0
ovn_cluster_state{ports="6643"} 0.0
ovn_cluster_state{ports="6644"} 0.0
# HELP ovn_cluster_state State of ovn ovn_cluster_state
# TYPE ovn_cluster_state gauge
ovn_cluster_state{certs="server.crt"} 0.0
ovn_cluster_state{certs="cluster.crt"} 1.0
```


docs/changelog.rst
**Changelog**

0.1
^^^
Jun 2023

* Created


prometheus-microovn-exporter/config.py
"""Configuration loader."""
import sys
from collections import OrderedDict
from logging import getLogger
from typing import Any, Dict, Union

import confuse


class ConfigMeta(type):
    """Singleton metaclass for the Config."""

    _instance = None

    def __new__(cls, *args: Any, **kwargs: Any) -> type:  # pylint: disable=C0204
        if not cls._instance:
            cls._instance = super(ConfigMeta, cls).__new__(cls, *args, **kwargs)
            cls._instance.config = confuse.Configuration("PrometheusMicroovnExporter", __name__)
        return cls._instance


class Config(metaclass=ConfigMeta):
    """Configuration class for PrometheusMicroovnExporter."""

    config: confuse.Configuration = None

    def __init__(self, args: Union[Dict, None] = None) -> None:
        """Initialize the config class."""
        if not self.config:
            self.config = confuse.Configuration("PrometheusMicroovnExporter", __name__)

        self.logger = getLogger(__name__)

        if args:
            self.config.set_args(args, dots=True)

        self.validate_config_options()

    def validate_config_options(self) -> None:
        """Validate the configuration values against a template."""
        template = {
            "exporter": OrderedDict(
                [
                    ("port", confuse.Choice(range(0, 65536), default=9999)),
                    ("collect_interval", int),
                ]
            ),
            "microovn_env": str,
            "microovn_cluster": OrderedDict([("path", str),]),
            # "microovn_services": OrderedDict(
            #     [
            #         ("path", str),
            #         ("central", str),
            #         ("chassis", str),
            #         ("switch", str),
            #         ("daemon", str),
            #     ]
            # ),
            "microovn_certs": OrderedDict(
                [
                    ("server", str),
                    ("cluster", str),
                ]
            ),
            "ovn_cluster": OrderedDict([("path", str),]),
            "ovn_certs": OrderedDict(
                [
                    ("server", str),
                    ("cluster", str),
                ]
            ),
            # "mode": confuse.Choice(("microovn", "ovn"), default="microovn"),
            "mode": str, 
            "debug": bool,
        }

        try:
            self.config.get(template)
            self.config["mode"].as_choice(["microovn", "ovn"])
            self.config["microovn_cluster"]['path'].as_filename()
            self.config["ovn_cluster"]['path'].as_filename()
            self.logger.info("Configuration parsed successfully")
        except (
            KeyError,
            confuse.ConfigTypeError,
            confuse.ConfigValueError,
            confuse.NotFoundError,
        ) as err:
            self.logger.error("Error parsing configuration values: %s", err)
            sys.exit(1)

    def get_config(self, section: Union[Dict, None] = None) -> confuse.Configuration:
        """Return the config."""
        if section:
            return self.config[section]

        return self.config


prometheus-microovn-exporter/exporter.py
"""Exporter module."""
from time import sleep
import sys
from logging import getLogger
from typing import Any, Dict

from prometheus_client import CollectorRegistry, Gauge, start_http_server, REGISTRY

from prometheus_microovn_exporter.collector import Collector
from prometheus_microovn_exporter.config import Config


class ExporterDaemon:
    """Core class of the exporter daemon."""

    def __init__(self) -> None:
        """Create new daemon and configure runtime environment."""
        self.config = Config().get_config()
        self.logger = getLogger(__name__)
        self.logger.info(f"Parsed config: {self.config.config_dir()}")
        self.collector = Collector()
        self.logger.debug("Exporter initialized")

    def collect(self) -> None:
        """Call Collector"""
        REGISTRY.register(self.collector)
        while True:
            try:
                self.logger.info("Collecting gauges...")
                sleep(self.config["exporter"]["collect_interval"].get(int) * 60)
            except Exception as exception:  # pylint: disable=W0703
                self.logger.error(f"Collection job resulted in error: {exception}")
                sys.exit(1)

    def run(self) -> None:
        """Run exporter"""
        self.logger.debug("Running prometheus client http server.")
        start_http_server(self.config["exporter"]["port"].get(int))
        try:
            self.collect()
        except KeyboardInterrupt as exception:
            # Gracefully handle keyboard interrupt
            self.logger.info(f"{exception}: Exiting...")
            sys.exit(0)


if __name__ == "__main__":
    obj = ExporterDaemon()
    obj.run()

prometheus-microovn-exporter/collector.py
"""Collector module."""
from logging import getLogger
from typing import Any, Dict
from prometheus_client.core import GaugeMetricFamily

from prometheus_microovn_exporter.config import Config
from prometheus_microovn_exporter.ovn_scraper import OvnScraper

class Collector:
    """Core class of the PrometheusMicroovnExporter collector."""

    def __init__(self) -> None:
        """Create new collector and configure runtime environment."""
        self.config = Config().get_config()
        self.logger = getLogger(__name__)
        self.data: Dict[str, Any] = {}
        self.mode = self.config["mode"].get(str)
        self.logger.debug("Collector initialized")
        self.ovn_scraper = OvnScraper()

    def create_gauge(self, element) -> GaugeMetricFamily:
        gauge_name = f"ovn_cluster_state"
        gauge_desc = f"State of ovn ovn_cluster_state"
        gauge = GaugeMetricFamily(gauge_name, gauge_desc, labels=[element])
        return gauge

    def collect(self) -> Any:
        """Get stats from current host"""
        data = self.ovn_scraper.get_stats()
        gauges = {i: self.create_gauge(i) for i in data.keys()}
        for elem, gauge in gauges.items():
            for k,v in data[elem].items():
                gauge.add_metric([str(k)], v)
            # self.logger.info(f"Gauge for {elem} collected and ready for exporting.")
            yield gauge


if __name__ == "__main__":
    from prometheus_client.core import REGISTRY
    from prometheus_client import start_http_server
    import time
    collector = Collector()
    start_http_server(9999)
    REGISTRY.register(Collector())
    while True:
        time.sleep(10)

prometheus-microovn-exporter/ovn_scraper.py
"""Collector module."""
from logging import getLogger
from typing import Any, Dict
import subprocess as sp
import socket as sock
from datetime import datetime
import OpenSSL as openssl

from prometheus_microovn_exporter.config import Config


class OvnScraper:
    """Core class of the PrometheusMicroovnExporter collector's data source."""

    def __init__(self) -> None:
        """Read configs and initialize scrape."""
        self.config = Config().get_config()
        self.logger = getLogger(__name__)
        self.mode = self.config["mode"].get(str)
        self.logger.debug("Ovn Scraper initialized")
        if self.mode == "microovn":
            self.ovs_appctl = "microovn.ovs-appctl"
        else:
            self.ovs_appctl = "ovs-appctl"
        self.cluster_name = {
            "nb": "OVN_Northbound",
            "sb": "OVN_Southbound",
        }

    def _condense_cluser_status(self, cluster_status_dict: Dict [str, Dict]) -> Dict [str, int]:
        '''
        Status of
        0: good
        1: bad
        '''
        condensed_dict = {}
        for cluster, cluster_info in cluster_status_dict.items():
            status = 0
            if cluster_info["vote"].lower() == "unknown":
                status = 1
            if cluster_info["role"].lower() not in ["leader", "follower"]:
                status = 1
            if cluster_info["status"].lower() != "cluster member":
                status = 1 
            condensed_dict[cluster] = status
        self.logger.debug(f"Condensed cluster status dict to {condensed_dict}")
        return condensed_dict

    def _parse_cluster_status_output(self, to_parse : Dict [str, Any]) -> Dict [str, Any]:
        """Read a ovn central's cluster/status ouput and extract information.
        
        :param Dict [str, Any] to_parse: dict of nb/sb cluster to it's cluster status
            output in plaintext
        """
        final_dict = {}
        to_consider = ["role" ,"leader" ,"vote", "address", "status"]
        for unit, cluster_status in to_parse.items():
            status_dict = {}
            for row in cluster_status.split('\n'):
                key = row.split(':')[0].lower()
                value = row.split(':')[-1]
                if key in to_consider:
                    if key == "address":
                        status_dict[key] = ':'.join(row.split(':')[2:])
                    else:
                        status_dict[key] = value.lstrip(' ')
            final_dict[unit] = status_dict
        self.logger.debug(f"From cluster status, parsed {final_dict}")
        return final_dict

    @staticmethod
    def _get_db_ctl(cluster: str) -> str:
        """Static function that renders nb/sb database ctl file name.
        
        :param str cluster: cluster name, either nb or sb
        """
        return f"ovn{cluster}_db.ctl"

    def check_cluster_status(self) -> Dict[str, Any]:
        """
        Queries the nb and sb cluster of the unit for its cluster status.
        Returns a dictionary mapping the cluster (nb/sb) to a dict of values
        representating the state of the cluster in various aspects.
        """
        output_to_parse = {}
        clusters = ["nb", "sb"]
        for cluster in clusters:
            config_key = f"{self.config['mode']}_cluster"
            path = self.config[config_key]["path"]
            path = f"{path}/{self._get_db_ctl(cluster)}"
            # @TODO: Add timeout when checking cluster status
            cmd = [self.ovs_appctl, "-t", path, "cluster/status", self.cluster_name[cluster]]
            try:
                self.logger.debug(f"Running {cmd}")
                output = sp.run(cmd, stdout=sp.PIPE, stderr=sp.DEVNULL)
                output_to_parse[cluster] = output.stdout.decode('utf-8')
            except Exception as exception:
                self.logger.warning(f"Could not query {cluster} cluster for status")
            continue
        return self._parse_cluster_status_output(output_to_parse)
    
    def get_cluster_status(self) -> Dict[str, Any]:
        cluster_status_dict = self.check_cluster_status()
        return self._condense_cluser_status(cluster_status_dict)

    # def _get_local_ip(self) -> str:
    #     '''
    #     Parses local ip either from
    #     env file, if using microovn
    #     cluster status, if using ovn
    #     '''
    #     ip = None
    #     if self.mode == "microovn":
    #         try:
    #             env_file = self.config["microovn_env"]
    #             with open(env_file, 'r') as f:
    #                 env_data = f.read()
    #         except Exception as exception:
    #             self.logger.warning(f"Could not read {env_file}")
    #             return None
    #         env_data = env_data.split('\n')
    #         for row in env_data:
    #             if row.split('=')[0].lower() == "ovn_local_ip":
    #                 ip = row.split('=')[1]
    #     elif self.mode == "ovn":
    #         cluster_info = self.check_cluster_status()
    #         ip = cluster_info['nb']["address"].split(':')[0]
    #     return ip
    def _get_local_ip(self) -> str:
        '''
        Parses local ip from cluster status output
        @TODO: microovn provides a env file, should try to get something similar
            from ovn
        '''
        ip = None
        cluster_info = self.check_cluster_status()
        if 'nb' in cluster_info:
            ip = cluster_info['nb']["address"].split(':')[0]
        else:
            self.logger.debug(cluster_info)
        return ip

    def get_ports(self) -> Dict[int, int]:
        '''
        Returns a dictionary of port to state mapping where a state value of
        0: port if OPEN
        1: port is CLOSED
        Other values: UNKNOWN
        '''
        port_state = {}
        local_ip = self._get_local_ip()
        if not local_ip:
            self.logger.warning("Could not get cluster status for local ip")
            return
        ports = {
            6641: {"ip": "127.0.0.1", "msg": "OVN Northbound OVSDB Server"},
            6642: {"ip": "127.0.0.1", "msg": "OVN Southbound OVSDB Server"},
            6643: {"ip": local_ip, "msg": "OVN NB RAFT Control Plane"},
            6644: {"ip": local_ip, "msg": "OVN SB RAFT Control Plane"},
        }
        if self.mode == "ovn":
            ports[16642] = {"ip": "127.0.0.1", "msg": "OVN misc port"}
        for p in ports:
            create_socket = sock.socket(sock.AF_INET, sock.SOCK_STREAM)
            create_socket.settimeout(5)
            result = create_socket.connect_ex((ports[p]["ip"], p))
            create_socket.close()
            port_state[p] = int(result)
            if port_state[p] == 1:
                self.logger.warning(f"Port {p} for {ports[p]['msg']} is NOT open")
            elif port_state[p] == 0:
                self.logger.debug(f"Port {p} for {ports[p]['msg']} is OPEN")
            else:
                self.logger.warning(f"Port {p} for {ports[p]['msg']} is an an UNKNOWN state")
        return port_state

    @staticmethod
    def _cleanedup_cert_name(cert_name: str) -> str:
        return str(cert_name.split("/")[-1])

    def get_certs(self) -> Dict[str, int]:
        '''
        Checks the validity of certs and returns a state value where
        0: Valid
        1: Invalid
        2: Valid, but only for a maximum of 30 days
        '''
        config_key = f"{self.mode}_certs"
        cert_validity = {
            self._cleanedup_cert_name(str(cert)): 1 for cert in self.config[config_key].values()
        }
        for cert in self.config[config_key]:
            cert = self.config[config_key][cert].get(str)
            cert_data = None
            x509_cert = None
            not_after_timestamp = None
            try:
                with open(cert, 'r') as f:
                    cert_data = f.read()
            except Exception as exception:
                # @TODO: Need to find a good way of determining if
                # node is either a central or chassis
                self.logger.warning(f"Could not find {cert}")
                continue
            try:
                x509_cert = openssl.crypto.load_certificate(openssl.crypto.FILETYPE_PEM, cert_data)
                self.logger.debug(f"Loaded x509 object of {cert}")
            except Exception as exception:
                self.logger.error(f"Could not use openssl to load {cert}")
                continue
            try:
                not_after_timestamp = x509_cert.get_notAfter().decode("utf-8")
            except Exception as exception:
                self.logger.error(f"Could not decode {cert}")
                continue
            # get the number of valid days left for the cert, compating the cert's
            # notAfter date to "today's" date
            not_after = datetime.strptime(not_after_timestamp, "%Y%m%d%H%M%S%z").date()
            now = datetime.now().date()
            num_days = not_after - now
            num_days = num_days.days
            self.logger.info(f"{cert} valid for {num_days} days, till {not_after}")
            if num_days > 30:
                cert_validity[self._cleanedup_cert_name(cert)] = 0
            elif num_days > 0 and num_days <= 30:
                cert_validity[self._cleanedup_cert_name(cert)] = 2
        return cert_validity

    def get_stats(self) -> Dict[str, Dict]:
        """Get stats from current ovn cluster unit"""
        stats = {
            "cluster_status": self.get_cluster_status(),
            "ports": self.get_ports(),
            "certs": self.get_certs(),
        }
        return stats


if __name__ == "__main__":
    scraper = OvnScraper()
    for k,v in scraper.get_stats().items():
        print(f"RET for {k}", v)